### ISSUE TITLE
Problem IPv6 in commit 3b74e1088af7b525880b386136c0330bbed377a0

### ISSUE THREAD

**Issue by Andlui9** [accel-ppp/accel-ppp#16]
I had a problem updating a server to the mentioned commit version.

I will leave below the data of the problem that I was able to collect, as well as my configuration.

Problem data:
```
GNU gdb (Debian 7.12-6) 7.12.0.20161007-git
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from accel-pppd...done.
(gdb) run
Starting program: /usr/sbin/accel-pppd -c /etc/accel-ppp/accel-ppp.conf -p /var/run/accel-ppp.pid
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff52f7700 (LWP 11297)]
[New Thread 0x7ffff36c4700 (LWP 11298)]
[New Thread 0x7ffff2ec3700 (LWP 11299)]
[New Thread 0x7ffff7fc9700 (LWP 11608)]
[New Thread 0x7ffff26c2700 (LWP 11609)]
[New Thread 0x7ffff25c1700 (LWP 11610)]
[New Thread 0x7ffff24c0700 (LWP 11611)]
[New Thread 0x7ffff23bf700 (LWP 11612)]
[New Thread 0x7ffff22be700 (LWP 11613)]
[New Thread 0x7ffff21bd700 (LWP 11614)]
[New Thread 0x7ffff20bc700 (LWP 11615)]
[New Thread 0x7ffff1fbb700 (LWP 11616)]
[New Thread 0x7ffff1eba700 (LWP 11617)]
[New Thread 0x7ffff1db9700 (LWP 11618)]
[New Thread 0x7ffff15b8700 (LWP 11619)]

Thread 5 "accel-pppd" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff7fc9700 (LWP 11608)]
pthread_spin_lock () at ../sysdeps/x86_64/nptl/pthread_spin_lock.S:22
22	../sysdeps/x86_64/nptl/pthread_spin_lock.S: Arquivo ou diretório não encontrado.
(gdb) bt full
#0  pthread_spin_lock () at ../sysdeps/x86_64/nptl/pthread_spin_lock.S:22
No locals.
#1  0x00007ffff44eab00 in get_dp (ses=0x7fffc4024150) at /usr/local/src/accel/accel-ppp/accel-pppd/extra/ipv6pool.c:312
        it = 0x55555578ada0 <ipdb_handlers>
        pool = 0x0
        start = 0x0
#2  0x000055555557a850 in ipdb_get_ipv6_prefix (ses=0x7fffc4024150) at /usr/local/src/accel/accel-ppp/accel-pppd/ipdb.c:62
        ipdb = 0x7ffff46ec140 <ipdb>
        it = 0x0
#3  0x00007ffff42e2c61 in dhcpv6_send_reply (req=0x7fffbc0862d0, pd=0x7fffe4039db0, code=2) at /usr/local/src/accel/accel-ppp/accel-pppd/ipv6/dhcpv6.c:352
        reply = 0x7fffbc0877d0
        opt = 0x7fffbc086560
        opt1 = 0x7fffbc088950
        opt2 = 0x7fffbc088910
        opt3 = 0x3e
        ia_na = 0x7fffbc0878ba
        ia_addr = 0x7fffbc08789e
        ia_prefix = 0x7fffbc086580
        a = 0x55555587f030
        addr = {__in6_u = {__u6_addr8 = "\000\000\000\000\000\000\000\000Y\320WUUU\000", __u6_addr16 = {0, 0, 0, 0, 53337, 21847, 21845, 0}, __u6_addr32 = {0, 0, 1431818329, 21845}}}
        ses = 0x7fffc4024150
        f = 1
        f1 = 219
        f2 = 0
#4  0x00007ffff42e3cf3 in dhcpv6_recv_solicit (req=0x7fffbc0862d0) at /usr/local/src/accel/accel-ppp/accel-pppd/ipv6/dhcpv6.c:634
        pd = 0x7fffe4039db0
#5  0x00007ffff42e41c6 in dhcpv6_recv_packet (pkt=0x7fffbc0862d0) at /usr/local/src/accel/accel-ppp/accel-pppd/ipv6/dhcpv6.c:756
No locals.
#6  0x00007ffff42e43b1 in dhcpv6_read (h=0x7fffe4039dd0) at /usr/local/src/accel/accel-ppp/accel-pppd/ipv6/dhcpv6.c:815
        pd = 0x7fffe4039db0
        ses = 0x7fffc4024150
        n = 132
        addr = {sin6_family = 10, sin6_port = 8706, sin6_flowinfo = 0, sin6_addr = {__in6_u = {__u6_addr8 = "\376\200\000\000\000\000\000\000\002\031F\377\376", <incomplete sequence \355\253>, 
              __u6_addr16 = {33022, 0, 0, 0, 6402, 65350, 13566, 44013}, __u6_addr32 = {33022, 0, 4282784002, 2884449534}}}, sin6_scope_id = 1552}
        len = 28
        pkt = 0x7fffbc0862d0
        buf = 0x7fffbc08b1a0 "\001\266lJ"
#7  0x00007ffff7bd267d in ctx_thread (ctx=0x7fffc4024438) at /usr/local/src/accel/accel-ppp/accel-pppd/triton/triton.c:252
        h = 0x7fffe4039e28
        t = 0x7ffff625ea3e <ipoe_ctx_switch>
        call = 0x7ffff7fc8d30
        tt = 140736481869944
        events = 1
#8  0x00007ffff7bd2438 in triton_thread (thread=0x5555560eed90) at /usr/local/src/accel/accel-ppp/accel-pppd/triton/triton.c:192
        set = {__val = {516, 0 <repeats 15 times>}}
        sig = 10
        need_free = 0
        stack = 0x0
#9  0x00007ffff77b04a4 in start_thread (arg=0x7ffff7fc9700) at pthread_create.c:456
        __res = <optimized out>
        pd = 0x7ffff7fc9700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140737353914112, 766347238913097883, 140737488347662, 140737488347663, 140737352863744, 3, -766329668675288933, -766330781321772901}, 
              mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
---Type <return> to continue, or q <return> to quit---
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#10 0x00007ffff6974d0f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
No locals.
```

My conf:
```
[modules]
pppoe
ipoe
radius
ippool
shaper
sigchld
log_file
ipv6_nd
ipv6pool
ipv6_dhcp
auth_pap
auth_chap_md5
auth_mschap_v2
auth_mschap_v1
pppd_compat
connlimit

[core]
thread-count=10
log-error=/var/log/accel-ppp/core.log

[common]
verbose=1
sid-case=lower
single-session=replace
sid-source=urandom
max-starting=100

[auth]
verbose=1
noauth=0
any-login=0

[ppp]
verbose=1
ccp=0
mtu=1480
max-mtu=1492
min-mtu=1452
max-mru=1492
mppe=deny
pcomp=deny
check-ip=1
accomp=deny
sid-case=lower
unit-cache=0
unit-preallocate=1
lcp-echo-timeout=60
lcp-echo-failure=3
lcp-echo-interval=20
single-session=replace
ipv4=require
ipv6=allow
ipv6-intf-id=0:0:0:1
ipv6-peer-intf-id=0:0:0:2
ipv6-accept-peer-intf-id=1

[pppoe]
verbose=1
mppe=deny
padi-limit=0
ip-pool=pool-pppoe-1
ac-name=accel
service-name=accel._____________net
accept-any-service=1
called-sid=ifname
sid-uppercase=1
ifname-in-sid=called-sid
mac-filter=/etc/accel-ppp/mac-blacklist,deny
pado-delay=0,1:500,2:750,3:1000

vlan-timeout=1296000
vlan-name=%I.%N

vlan-mon=enp8s0f1,1000-1099,2000-2032
interface=enp8s0f1
interface=re:^enp8s0f1\.(10[0-8][0-9]|109[0-9]|20[0-2][0-9]|203[0-2])$

[ipoe]
verbose=1
username=lua:username
password=username
renew-time=120
lease-time=180
max-lease-time=240
unit-cache=0
proxy-arp=1
attr-dhcp-client-ip=Framed-IP-Address
attr-dhcp-router-ip=DHCP-Router-IP-Address
lua-file=/etc/accel-ppp/ipoe-accel-ppp.lua
ip-pool=pool-ipoe-1
check-mac-change=1
soft-terminate=1
calling-sid=mac

vlan-timeout=1296000
vlan-name=%I.%N

vlan-mon=enp8s0f1,2100-2299
interface=enp8s0f1,mode=L2,shared=1,start=dhcpv4,ifcfg=1,proxy-arp=2,ipv6=1
interface=re:^enp8s0f1\.(21[0-9]{2}|22[0-8][0-9]|229[0-9])$,mode=L2,shared=1,start=dhcpv4,ifcfg=1,proxy-arp=2,ipv6=1

gw-ip-address=100.101.32.1/24
gw-ip-address=100.101.33.1/24
gw-ip-address=100.101.34.1/24
gw-ip-address=100.101.35.1/24
gw-ip-address=100.101.36.1/24
gw-ip-address=100.101.37.1/24
gw-ip-address=100.101.38.1/24
gw-ip-address=100.101.39.1/24
gw-ip-address=100.102.32.1/24
gw-ip-address=198.18.196.1/22
gw-ip-address=_______.19.6

[dns]
dns1=_______.19.36
dns2=_______.19.37

[radius]
verbose=1
timeout=3
max-try=0
sid_in_auth=1
acct-timeout=0
interim-verbose=0
nas-identifier=accel
acct-interim-interval=600
gw-ip-address=_______.19.6
nas-ip-address=_______.19.6
dae-server=0.0.0.0:3799,_____________
dictionary=/usr/share/accel-ppp/radius/dictionary
server=_____________,_____________,auth-port=3112,acct-port=0,req-limit=0,fail-timeout=0,max-fail=0,weight=1
server=_____________,_____________,auth-port=0,acct-port=3113,req-limit=0,fail-timeout=0,max-fail=0,weight=1

[client-ip-range]
disable

[ip-pool]
verbose=1
shuffle=0
attr=Framed-Pool
gw-ip-address=_______.19.6

100.100.36.0/22,name=pool-pppoe-2
100.100.32.0/22,name=pool-pppoe-1,next=pool-pppoe-2

100.101.39.2-254,name=pool-ipoe-8
100.101.38.2-254,name=pool-ipoe-7,next=pool-ipoe-8
100.101.37.2-254,name=pool-ipoe-6,next=pool-ipoe-7
100.101.36.2-254,name=pool-ipoe-5,next=pool-ipoe-6
100.101.35.2-254,name=pool-ipoe-4,next=pool-ipoe-5
100.101.34.2-254,name=pool-ipoe-3,next=pool-ipoe-4
100.101.33.2-254,name=pool-ipoe-2,next=pool-ipoe-3
100.101.32.2-254,name=pool-ipoe-1,next=pool-ipoe-2

198.18.196.0/22,name=pgcorte

[ipv6-dns]
#_____:_____:a:fa::36
#_____:_____:a:fa::37

2001:4860:4860::8888
2001:4860:4860::8844

[ipv6-nd]
AdvAutonomousFlag=1

[ipv6-pool]
_____:_____:e::/48,64
delegate=_____:_____:f::/48,60

[ipv6-dhcp]
verbose=1
pref-lifetime=604800
valid-lifetime=2592000
route-via-gw=1

[log]
copy=1
color=1
level=3
log-debug=/var/log/accel-ppp/debug.log
log-file=/var/log/accel-ppp/accel-ppp.log
log-emerg=/var/log/accel-ppp/emerg.log
log-fail-file=/var/log/accel-ppp/auth-fail.log

[pppd-compat]
verbose=1
radattr-prefix=/var/run/radattr
ip-up=/usr/local/bin/pppoe-up.sh
ip-down=/usr/local/bin/pppoe-down.sh
ip-change=/usr/local/bin/pppoe-change.sh

[shaper]
verbose=1
ifb=ifb0
attr=Filter-Id

up-limiter=police
down-limiter=tbf

#up-limiter=htb
#down-limiter=htb
#leaf-qdisc=sfq perturb 10
#leaf-qdisc=fq_codel limit 512 flows 1024 quantum 1480 target 30ms interval 6ms noecn

[cli]
verbose=0
prompt=accel
tcp=0.0.0.0:2001
telnet=0.0.0.0:2000
sessions-columns=ifname,username,ip,rate-limit,called-sid,calling-sid,sid,tx-bytes,rx-bytes,state,uptime

[connlimit]
verbose=1
burst=0
limit=60/min   
timeout=5
```

**Comment by Andlui9** [comment]
I needed to say the version and flavor of my linux, Debian 9 fully updated, with kernel 4.9.0-15-amd64.

**Comment by Andlui9** [comment]
I also realized that if the radius does not respond it has problems ...

```
(gdb) run
Starting program: /usr/sbin/accel-pppd -c /etc/accel-ppp/accel-ppp.conf -p /var/run/accel-ppp.pid
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff52f7700 (LWP 34883)]
[New Thread 0x7ffff36c4700 (LWP 34884)]
[New Thread 0x7ffff2ec3700 (LWP 34885)]
[New Thread 0x7ffff7fc9700 (LWP 34888)]
[New Thread 0x7ffff26c2700 (LWP 34889)]
[New Thread 0x7ffff25c1700 (LWP 34890)]
[New Thread 0x7ffff24c0700 (LWP 34891)]
[New Thread 0x7ffff23bf700 (LWP 34892)]
[New Thread 0x7ffff22be700 (LWP 34893)]
[New Thread 0x7ffff21bd700 (LWP 34894)]
[New Thread 0x7ffff20bc700 (LWP 34895)]
[New Thread 0x7ffff1fbb700 (LWP 34896)]
[New Thread 0x7ffff1eba700 (LWP 34897)]
[New Thread 0x7ffff1db9700 (LWP 34898)]
[New Thread 0x7ffff15b8700 (LWP 34899)]

Thread 13 "accel-pppd" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff1fbb700 (LWP 34896)]
0x00007ffff5912c3b in rad_auth_finalize (rpd=0x7fffe800bf48, r=1) at /usr/local/src/accel/accel-ppp/accel-pppd/radius/auth.c:158
158                     if (rpd->auth_ctx->req) rad_req_free(rpd->auth_ctx->req);
(gdb) bt full
#0  0x00007ffff5912c3b in rad_auth_finalize (rpd=0x7fffe800bf48, r=1) at /usr/local/src/accel/accel-ppp/accel-pppd/radius/auth.c:158
No locals.
#1  0x00007ffff5912f23 in rad_auth_timeout (t=0x7fffe800c278) at /usr/local/src/accel/accel-ppp/accel-pppd/radius/auth.c:216
        req = 0x7fffe800c248
#2  0x00007ffff7bd25a7 in ctx_thread (ctx=0x7fffec0028a8) at /usr/local/src/accel/accel-ppp/accel-pppd/triton/triton.c:234
        h = 0x7ffff6476a60 <arp_ctx>
        t = 0x7fffe800cc98
        call = 0x7ffff1fbad30
        tt = 1
        events = 32767
#3  0x00007ffff7bd2438 in triton_thread (thread=0x5555560e11a0) at /usr/local/src/accel/accel-ppp/accel-pppd/triton/triton.c:192
        set = {__val = {516, 0 <repeats 15 times>}}
        sig = 10
        need_free = 0
        stack = 0x0
#4  0x00007ffff77b04a4 in start_thread (arg=0x7ffff1fbb700) at pthread_create.c:456
        __res = <optimized out>
        pd = 0x7ffff1fbb700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140737253193472, 8080101218668412613, 140737488347662, 140737488347663, 140737252143104, 3, -8080114376768586043, -8080117672299954491}, 
              mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#5  0x00007ffff6974d0f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
No locals.
```

**Comment by Andlui9** [comment]
In my tests, I realized that the problem is related to the fact that there are no parameters in the configuration of the ip pool session, the parameters attr-prefix=Delegated-IPv6-Prefix-Pool and attr-address=Stateful-IPv6-Address-Pool, and when my radius sent these attributes, the accel presented the problem. I could not validate that the same happens for IPv4, but it is worth investigating the code.

That's what I realized.
